conduit.go

```diff
@@ -0,0 +1,331 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+)
+
+var (
+	LocalPort int64
+)
+
+const tunnelInfo = `
+The following services are ready for you to connect to:
+
+{{range $serviceName, $services := .VcapServices}}
+	{{range $serviceIndex, $service := $services}}
+		service: {{$service.Name}} ({{$serviceName}})
+		host: {{$service.Credentials.Host}}
+		port: {{$service.Credentials.Port}}
+		username: {{$service.Credentials.Username}}
+		password: {{$service.Credentials.Password}}
+		db: {{$service.Credentials.Name}}
+	{{end}}
+{{end}}
+`
+
+func init() {
+	ConnectService.Flags().Int64VarP(&LocalPort, "local-port", "p", 7080, "start selecting local ports from")
+}
+
+func waitForConnection(addr string) chan error {
+	timeout := 3 * time.Second
+	connection := make(chan error)
+	go func() {
+		defer close(connection)
+		tries := 0
+		for {
+			if tries > 5 {
+				time.Sleep(2 * time.Second)
+			} else {
+				time.Sleep(1 * time.Second)
+			}
+			tries++
+			debug("waiting for", addr, "attempt", tries)
+			conn, err := net.DialTimeout("tcp", addr, timeout)
+			if err != nil {
+				if tries < 15 {
+					continue
+				}
+				connection <- fmt.Errorf("connection fail after %d attempts: %s", tries, err)
+				break
+			}
+			defer conn.Close()
+			connection <- nil
+			break
+		}
+	}()
+	return connection
+}
+
+var ConnectService = &cobra.Command{
+	Use: "conduit [flags] SERVICE_INSTANCE [-- COMMAND]",
+	Example: `  Create a tunnel between your machine and a remote running service:
+  cf conduit my-service
+  
+  Run a local application that uses VCAP_SERVICES environment to connect to remote services:
+  cf conduit my-service-1 my-service-2 -- /path/to/app
+    
+  Export a postgres database:
+  cf conduit postgres-instance -- pg_dump -f backup.sql
+
+  Import a postgres script:
+  cf conduit postgres-instance -- psql < backup.sql
+
+  Dump a mysql database:
+  cf conduit mysql-instance -- mysqldump --all-databases > backup.sql
+
+  Import a mysql script:
+  cf conduit mysql-instance -- psql < backup.psql
+  `,
+	Short: "enables temporarily binding services to local running processes",
+	Long:  "spawns a temporary application, binds your desired service and creates an ssh tunnel from the application to your local machine enabling communication directly with the remote service.",
+	Args: func(cmd *cobra.Command, args []string) error {
+		if cmd.ArgsLenAtDash() > -1 {
+			if cmd.ArgsLenAtDash() < 1 {
+				return errors.New("requires at least one SERVICE_INSTANCE argument to be specified")
+			}
+		} else if len(args) < 1 {
+			return errors.New("requires at least one SERVICE_INSTANCE argument to be specified")
+		}
+		return nil
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		// create status writer
+		status := NewStatus(os.Stderr)
+		defer status.Done()
+		// parse args
```

Comment:
Could we extract this sort of thing to a function, instead of delimiting with a comment?

Reply:
I want to say yes, but the code was actually factored _back_ to a single giant function from many smaller ones since it was easier to follow due to the script-like nature of the task.

I agree I'm not happy with the comment delimited code but in the absence of tests to ensure intension, I was using them as somewhat of an implementation guide to help with the _next_ refactor.

I did try to break stuff up like:

```
ensureEach([]errors{
    status.WithMessage("doing thing", func(){
        // do thing
        return nil
    }),
status.WithMessage("doing thing", func(){
    // do thing
    return nil
})
```

... but I think that just made things worse , so I'm going to leave as is for now

Reply:
:+1: